clients/src/main/java/org/apache/kafka/clients/ApiVersions.java

```diff
@@ -30,18 +30,29 @@
  * NOTE: This class is intended for INTERNAL usage only within Kafka.
  */
 public class ApiVersions {
-
     private final Map<String, NodeApiVersions> nodeApiVersions = new HashMap<>();
     private byte maxUsableProduceMagic = RecordBatch.CURRENT_MAGIC_VALUE;
+    private ChangeListener changeListener = null;
+
+    public ApiVersions() {
+    }
+
+    public ApiVersions(ChangeListener changeListener) {
+        this.changeListener = changeListener;
+    }
 
     public synchronized void update(String nodeId, NodeApiVersions nodeApiVersions) {
         this.nodeApiVersions.put(nodeId, nodeApiVersions);
         this.maxUsableProduceMagic = computeMaxUsableProduceMagic();
```

Comment:
It'd be ideal to move this logic to a listener implementation as well but maxUsableProduceMagic() is invoked from KafkaProducer, Sender and RecordAccumulator making it harder to encapsulate in a single layer. We could potentially make this change in a follow-on PR.